core/src/main/java/io/grpc/internal/DelayedClientCall.java

```diff
@@ -0,0 +1,518 @@
+   * call.
+   */
+  @Nullable
+  private final ScheduledFuture<?> initialDeadlineMonitor;
+  private final Executor callExecutor;
+  private final Context context;
+  /** {@code true} once realCall is valid and all pending calls have been drained. */
+  private volatile boolean passThrough;
+  /**
+   * Non-{@code null} iff start has been called. Used to assert methods are called in appropriate
+   * order, but also used if an error occurs before {@code realCall} is set.
+   */
+  private Listener<RespT> listener;
+  // Must hold {@code this} lock when setting.
+  private ClientCall<ReqT, RespT> realCall;
+  @GuardedBy("this")
+  private Status error;
+  @GuardedBy("this")
+  private List<Runnable> pendingRunnables = new ArrayList<>();
+  @GuardedBy("this")
+  private DelayedListener<RespT> delayedListener;
+
+  DelayedClientCall(
+      Executor callExecutor, ScheduledExecutorService scheduler, @Nullable Deadline deadline) {
+    this.callExecutor = checkNotNull(callExecutor, "callExecutor");
+    checkNotNull(scheduler, "scheduler");
+    context = Context.current();
+    Deadline contextDeadline = context.getDeadline();
+    if (deadline == null && contextDeadline == null) {
+      this.initialDeadlineMonitor = null;
+      return;
+    }
+    long remainingNanos = Long.MAX_VALUE;
+    if (deadline != null) {
+      remainingNanos = Math.min(remainingNanos, deadline.timeRemaining(NANOSECONDS));
+    }
+    if (contextDeadline != null && contextDeadline.timeRemaining(NANOSECONDS) < remainingNanos) {
+      remainingNanos = contextDeadline.timeRemaining(NANOSECONDS);
+      if (logger.isLoggable(Level.FINE)) {
+        StringBuilder builder =
+            new StringBuilder(
+                String.format(
+                    "Call timeout set to '%d' ns, due to context deadline.", remainingNanos));
+        if (deadline == null) {
+          builder.append(" Explicit call timeout was not set.");
+        } else {
+          long callTimeout = deadline.timeRemaining(TimeUnit.NANOSECONDS);
+          builder.append(String.format(" Explicit call timeout was '%d' ns.", callTimeout));
+        }
+        logger.fine(builder.toString());
+      }
+    }
+    long seconds = Math.abs(remainingNanos) / TimeUnit.SECONDS.toNanos(1);
+    long nanos = Math.abs(remainingNanos) % TimeUnit.SECONDS.toNanos(1);
+    final StringBuilder buf = new StringBuilder();
+    if (remainingNanos < 0) {
+      buf.append("ClientCall started after deadline exceeded. Deadline exceeded after -");
+    } else {
+      buf.append("Deadline exceeded after ");
+    }
+    buf.append(seconds);
+    buf.append(String.format(".%09d", nanos));
+    buf.append("s. ");
+    /** Cancels the call if deadline exceeded prior to the real call being set. */
+    class DeadlineExceededRunnable implements Runnable {
+      @Override
+      public void run() {
+        synchronized (DelayedClientCall.this) {
+          if (realCall != null) {
```

Comment:
It's unclear to me how much this does. What would happen if we didn't have this check?

I feel like we either don't need this check, or we need to pass a boolean to cancel(Status) saying not to cancel if realCall is set, in a non-racy way.

Reply:
The interop test (AbstractInteropTest.deadlineInPast()) was very flaky without it. So I guessed it was due to racy cancellation. Adding it just fixed the flakiness. Let me check if this is the right fix...

Reply:
Discussed "offline." This is legitimate because ClientCall.cancel() is non-thread-safe and will cause exceptions if you try to send after it. That is different from in the stream where cancel() is thread-safe and wouldn't trigger any checkState()s. But as-is it isn't atomic; we need to move the `realCall != null` check into the private cancel() function to make it atomic.

Reply:
Done.